File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = float('inf')
    n = len(diffs)
    left = 1
    right = max(diffs)
    
    def is_possible(level):
        total = 0
        for i in range(n):
            if diffs[i] <= level:
                total += times[i]
            else:
                total += (diffs[i] - level) * (times[i-1] + times[i]) + times[i]
            
            if total > limit:
                return False
        
        return True
    
    while left <= right:
        mid = (left + right) // 2
        
        if is_possible(mid):
            answer = min(mid, answer)
            right = mid - 1
        else:
            left = mid + 1
    
    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public boolean checkIsPossible(int[] diffs, int[] times, long limit, int level){
        long requireTime = 0;
        int n = diffs.length;
        for(int i = 0; i < n; i++){
            if(diffs[i] <= level){
                requireTime += times[i];
            }
            else if(diffs[i] > level){
                requireTime += (diffs[i] - level) * (times[i-1] + times[i]) + times[i];
            }
        }

        return requireTime <= limit ? true : false;
    }

    public int binarySearch(int[] diffs, int[] times, long limit){
        int left = 1;
        int right = Arrays.stream(diffs).max().getAsInt();
        int minimum = right;
        while(left <= right){
            int mid = (left + right) / 2;
            if(checkIsPossible(diffs, times, limit, mid)){
                minimum = Math.min(right,mid);
                right = mid-1;
            }
            else{
                left = mid+1;
            }   
        }
        
        return minimum;
    }
    
    public int solution(int[] diffs, int[] times, long limit) {
        return binarySearch(diffs, times, limit);
        
    }
}